R/axis_handling.R

```diff
@@ -0,0 +1,151 @@
+#'  be abbreviated. This defaults to \code{c(1)}, which retains the original
+#'  first label and abbreviates subsequent ones. If all breaks should be
+#'  abbreviated, this can be set to NULL.
+#'@param full_by_year Vector of integers, the value of breaks that should not be
+#'  abbreviated. Defaults to NULL.
+#'
+#'@describeIn abbr_years_cont For continuous scales
+#'@export
+abbr_years_cont <- function(full_by_pos = c(1),
+                            full_by_year = NULL) {
+
+  fxn <- function(breaks) {
+
+    # Determine length of each break. Note that when breaks have been modified,
+    # ggplot sometimes adds an additional NA break in the first and last
+    # positions. We check whether this is the case and, if so, remove those for
+    # testing purposes.
+    lengths <- stringr::str_length(as.integer(breaks[which(!is.na(breaks))]))
+
+    # # Stop if the breaks are not in a four-digit format.
+    # if (!(min(lengths == 4) & max(lengths == 4))) {
+    #   stop("Breaks are not in a four-digit format. Cannot abbreviate.")
+    # }
+
+    # # Stop if breaks cannot be coerced to a number
+    # tryCatch({as.integer(breaks)},
+    #          warning = function(w){stop("Breaks cannot be coerced to a year. Cannot abbreviate.")})
+
+    # Abbreviate all values
+    abbr <- paste0("'",substr(breaks,3,4))
+
+    # Account for leading NAs and increment up positions accordingly
+    leading_na <- which.min(is.na(breaks)) - 1
+    full_by_pos <- full_by_pos + leading_na
+
+    # Now convert referenced dates into positions
+    if(!is.null(full_by_year)) {
+      full_by_pos <- sort(unique(c(full_by_pos,match(full_by_year,breaks))))
+    }
+
+    # Add back full years for specified positions
+    if(!is.null(full_by_pos)) {
+      abbr[full_by_pos] <- breaks[full_by_pos]
+    }
+
+    return(abbr)
+  }
+
+  return(fxn)
+}
+
+#'@describeIn abbr_years_cont For date scales
+#'
+#'@export
+abbr_years_date <- function(full_by_pos = c(1),
+                            full_by_year = NULL) {
+  fxn <- function(breaks) {
+
+  # Convert given year integers into the number of days since 1970 (note this is
+    # how the breaks are handled internally by ggplot - they live in the following
+    # location of a ggplot chart 'a' that has been passed to ggplot_build()
+    #          ggplot_build(a)$layout$panel_params[[1]]$x$breaks
+    # Note that breaks for the y axis can be accessed similarly.
+    if (!is.null(full_by_year))
+      {full_by_year <- as.numeric(as.Date(paste0(full_by_year,"-01-01")))}
+
+    # Stop if breaks cannot be coerced to a date
+    tryCatch({as.Date(breaks)},
+             warning = function(w){stop("Breaks cannot be coerced to a date. Cannot abbreviate.")})
+
+    # Abbreviate all values - ggplot appears to correctly use the names of the
+    # breaks (which are the year values) when performing this operation on the
+    # breaks.
+    abbr <- paste0("'",substr(breaks,3,4))
```

Comment:
it's weird that this works with `breaks` but below you need `names(breaks)`. I wonder, if you combine the two functions into one, whether it could be easiest and safer to do something at the top like:

```
# not sure these semantics are right:
if(dateaxis){breaks <- year(as.date(breaks))}
```

and then let the rest of the function work identically?

Reply:
Yeah I will play around with that. Some of the internal mechanics are opaque though, so it might stay as is.

Reply:
I tried this, and was not able to get it to work. I could have been doing something wrong, but couldn't figure it out. Decided to stop as this is a rather weird rabbit hole and it seems to be working fine as is right now.